plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,406 @@
+            Let (a, usBody) NonRec bs letBody
+                -- The binding can be placed inside a `Let`, if the right hand sides of the
+                -- bindings of the `Let` do not mention `var`.
+                | (var ^. PLC.unique) `Set.notMember` foldMap bindingUniqs bs ->
+                    Let (a, Set.union usBind usBody) NonRec bs <$> go b letBody
+                | (var ^. PLC.unique) `Set.notMember` termUniqs letBody -> do
+                    splitBindings (var ^. PLC.unique) (NonEmpty.toList bs) >>= \case
+                        Just
+                            ( before
+                                , TermBind (a', usBind') s' var' rhs'
+                                , after
+                                , inManyOccRhs
+                                ) -> do
+                                -- `letBody` does not mention `var`, and there is exactly one
+                                -- RHS in `bs` that mentions `var`, so we can place `b`
+                                -- inside one of the RHSs in `bs`.
+                                b'' <-
+                                    TermBind (a', Set.union usBind usBind') s' var'
+                                        <$> local
+                                            (over ctxtInManyOccRhs (|| inManyOccRhs))
+                                            (go b rhs')
+                                let bs' = NonEmpty.appendr before (b'' :| after)
+                                pure $ Let (a, Set.union usBind usBody) NonRec bs' letBody
+                        _ -> giveup
+            IWrap (a, usBody) ty1 ty2 iwrapBody ->
+                -- A binding can always be placed inside an `IWrap`.
+                IWrap (a, Set.union usBind usBody) ty1 ty2 <$> go b iwrapBody
+            Unwrap (a, usBody) unwrapBody ->
+                -- A binding can always be placed inside an `Unwrap`.
+                Unwrap (a, Set.union usBind usBody) <$> go b unwrapBody
+            _ -> giveup
+        -- TODO: implement float-in for type and datatype bindings.
+        _ -> giveup
+      where
+        giveup =
+            let us = Set.union (termUniqs body) (bindingUniqs b)
+             in pure $ Let (letAnn, us) NonRec (pure b) body
+
+{- | Split the given list of bindings, if possible.
+ If the input contains exactly one binding @b@ that mentions the given `PLC.TermUnique`,
+ return @Just (before_b, b, after_b)@.
+ Otherwise, return `Nothing`.
+-}
+splitBindings ::
+    PLC.HasUnique name PLC.TermUnique =>
+    PLC.TermUnique ->
+    [Binding tyname name uni fun (a, Uniques)] ->
+    Reader
+        FloatInContext
+        ( Maybe
+            ( [Binding tyname name uni fun (a, Uniques)]
+            , Binding tyname name uni fun (a, Uniques)
+            , [Binding tyname name uni fun (a, Uniques)]
+            , Bool
+            -- \^ Whether the chosen binding occurs more than once.
+            )
+        )
+splitBindings u bs = case is of
```

Comment:
This function merges two things:
1. Identifying the sole binding that uses the variable
2. Determining whether that binding is safe to float into

I think it would be clearer if we just did 1 and then we did 2 at the call site. We already handle the rest of the floating safety stuff in `go`.

Also as it is the Haddock is misleading.

Reply:
Good point!